### ISSUE TITLE
[Flang] Incorrect diagnostic on intrinsic `ASSOCIATED` when the `TARGET` argument is unlimited polymorphic

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#125774]
Consider the following code
```
    type :: bType
        integer*4 :: i1
    end type

    type base
        type (bType), pointer :: b1
    end type

    class (*), allocatable, target :: x

    type (base) :: b1

    if (.not. associated (b1%b1, x)) error stop(2_4)
end

```

Flang currently issues an error as:
```
./t.f:13:15: error: Arguments of ASSOCIATED() must be a pointer and an optional valid target
      if (.not. associated (b1%b1, x)) error stop(2_4)
                ^^^^^^^^^^^^^^^^^^^^^
```

The code seems valid as the standard requires the type of `TARGET` to be compatible with `POINTER` argument.
Unlimited polymorphic entity is type compatible with all entities.

All ifort, gfortran and XLF compile the code successfully.

**Comment by klausler** [comment]
Would `b1%b1 => x` be a valid pointer assignment statement?

**Comment by klausler** [comment]
```
type t                                                                                                                                                          
end type                                                                                                                                                     
real, pointer :: rptr                                                                                                                                           
type(t), pointer :: tptr                                                                                                                                
class(*), pointer :: ulpp  
! nvfortran and cray emit no error for any of these                                                                                                                                     
print *, associated(rptr, ulpp) ! error with gfortran, ifx, nag; warning with xlf                                                                               
print *, associated(ulpp, rptr) ! error with ifx, nag; warning with xlf                                                                                         
print *, associated(tptr, ulpp) ! error with gfortran, nag; warning with xlf                                                                                    
print *, associated(ulpp, tptr) ! ok with all                                                                                                                   
print *, associated(rptr, tptr) ! not standard; error with gfortran, ifx, nag, warning with xlf                     
print *, associated(tptr, rptr) ! not standard; error with gfortran, ifx, nag, warning with xlf                                                               
end                                                                                                                                                             
```

I think that only the last two calls should be flagged as errors.  I might emit a portability warning for the first three cases.
